### ISSUE TITLE
Windows DX: Window is not centred on startup

### ISSUE THREAD

**Issue by MichaelDePiazzi** [MonoGame/MonoGame#4740]
When using Windows DirectX, the game window is no longer centred on start up. Windows GL still works correctly.

Does anyone have any ideas on what might have broken this? If not, I will try take a look into this later when I have time.

**Comment by tomspilman** [comment]
My guess is this commit maybe...

 https://github.com/mono/MonoGame/pull/4425

Specifically could be this change...

https://github.com/mono/MonoGame/blob/develop/MonoGame.Framework/Windows/WinFormsGameWindow.cs#L139

I don't think we ever explicitly called any API to center the form window, but maybe we should right after the `Show`.

Also it would be nice to work out a unit test to verify this isn't broken again.  Maybe we can abuse the `GameWindow.Position` and `GameWindow.ClientBounds` to somehow verify this.

**Comment by tomspilman** [comment]
Actually just fired up a test i was working on and it seems the WinDX window is centered correctly for me with the default window sizing.

Are you changing the back buffer size in the ctor or during initialize?

**Comment by MichaelDePiazzi** [comment]
@tomspilman I'm setting the back buffer size in the constructor.

**Comment by Tzenchor** [comment]
I have done a little test in my game and it does not matter if the back buffer is set in the `Constructor` or in `Initialize`, the window position is the same. But when set on `Initialize` you can see the window resizing

**Comment by MichaelDePiazzi** [comment]
@tomspilman Actually, when using the default window sizing it is centred correctly. If you change the window size, then it's not centred properly. So the centring now seems to be based from the default window size instead of the specified window size.

**Comment by tomspilman** [comment]
I figured out the reason for this.

Make the following change to WinFormsGameWindow:

```
        public override Rectangle ClientBounds
        {
            get
            {
                var position = Point.Empty;
                if (_form.Visible)
                    position = _form.PointToScreen(position);
                var size = _form.ClientSize;
                return new Rectangle(position.X, position.Y, size.Width, size.Height);
            }
        }
```

If a call to get the client window came in before the form was visible then the `PointToScreen` would force the window to calculate its start position.  Later you change the window size and since the start position has already been set... it doesn't get fixed to center on show.

**Comment by tomspilman** [comment]
Now that doesn't seem like the right fix... XNA will return the window position even when the window isn't visible yet.  So we really need to change our centering technique to not depend on the start position.

**Comment by tomspilman** [comment]
So another broken thing in this area.

In XNA if you change the back buffer size on a windowed game and you haven't moved the window yet...  the game window will resize and stay centered.  If you have moved the window once it won't center the window... it will just change the size.

Under MonoGame we never recenter on a resize.

Minor difference, but a difference.

**Comment by MichaelDePiazzi** [comment]
@tomspilman Hmmm... does that maybe mean if the form's `ClientSize` is set correctly before retrieving `GameWindow.ClientBounds` that it would also fix this? (I'm wondering if this is this possibly why it worked before)

And that's interesting - I didn't realise that XNA would re-centre the window so long as you hadn't moved it yet.

**Comment by tomspilman** [comment]
> does that maybe mean if the form's `ClientSize` is set correctly before retrieving `GameWindow.ClientBounds` that it would also fix this?

No the bug was always there...  but appeared once internally we called `GameWindow.ClientBounds` during initialization.  So that is what exposed it.

If you can give me a day or two i'll setup a full set of unit tests and fix all this mess once and for all.

**Comment by MichaelDePiazzi** [comment]
@tomspilman No problem at all mate. Sounds like you have a better handle on how to fix this than I do. Let me know if you need anything from me.

**Comment by MichaelDePiazzi** [comment]
@tomspilman Have you possibly had a chance to look at this one yet?

**Comment by tomspilman** [comment]
Not yet... got pulled into a dozen other issues.  I'll try to get to this later this week.

**Comment by MichaelDePiazzi** [comment]
No problem, thanks Tom.

**Comment by MichaelDePiazzi** [comment]
Nice! I didn't realise PR #5139 fixed this. Just tested it and it's working now. Thanks @Jjagg! :)

**Comment by Jjagg** [comment]
You're welcome :)
 I tried a couple things, but eventually just computed the right position manually.

**Comment by EnemyArea** [comment]
this is'nt still fixed in the 3.6 update...

**Comment by MichaelDePiazzi** [comment]
Yeah, you're right. I just tried this myself with the latest source, and it's not working anymore. It must have been broken at some point...

I can't seem to re-open this issue for some reason, so I'll raise a new one...
